drivers/counter/counter_ifx_cat1.c

```diff
@@ -0,0 +1,537 @@
+/*
+ * Copyright (c) 2023 Cypress Semiconductor Corporation (an Infineon company) or
+ * an affiliate of Cypress Semiconductor Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief Counter driver for Infineon CAT1 MCU family.
+ */
+
+#define DT_DRV_COMPAT infineon_cat1_counter
+
+#include <zephyr/drivers/counter.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <cyhal_timer.h>
+#include <cyhal_gpio.h>
+
+#include <cyhal_tcpwm_common.h>
+#include <zephyr/dt-bindings/pinctrl/ifx_cat1-pinctrl.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ifx_cat1_counter, CONFIG_COUNTER_LOG_LEVEL);
+
+/* Device config structure */
+struct ifx_cat1_counter_config {
+	struct counter_config_info counter_info;
+	TCPWM_CNT_Type *reg_addr;
+	cyhal_gpio_t external_pin;
+	IRQn_Type irqn;
+	uint8_t irq_priority;
+};
+
+/* Data structure */
+struct ifx_cat1_counter_data {
+	cyhal_timer_t counter_obj;
+	cyhal_timer_cfg_t counter_cfg;
+	struct counter_alarm_cfg alarm_cfg_counter;
+	struct counter_top_cfg top_value_cfg_counter;
+	uint32_t guard_period;
+	cyhal_resource_inst_t hw_resource;
+	cyhal_source_t signal_source;
+	bool alarm_irq_flag;
+};
+
+/* Default Counter configuration structure */
+static const cy_stc_tcpwm_counter_config_t cyhal_timer_default_config = {
+	.period = 32768,
+	.clockPrescaler = CY_TCPWM_COUNTER_PRESCALER_DIVBY_1,
+	.runMode = CY_TCPWM_COUNTER_CONTINUOUS,
+	.countDirection = CY_TCPWM_COUNTER_COUNT_UP,
+	.compareOrCapture = CY_TCPWM_COUNTER_MODE_CAPTURE,
+	.compare0 = 16384,
+	.compare1 = 16384,
+	.enableCompareSwap = false,
+	.interruptSources = CY_TCPWM_INT_NONE,
+	.captureInputMode = 0x3U,
+	.captureInput = CY_TCPWM_INPUT_0,
+	.reloadInputMode = 0x3U,
+	.reloadInput = CY_TCPWM_INPUT_0,
+	.startInputMode = 0x3U,
+	.startInput = CY_TCPWM_INPUT_0,
+	.stopInputMode = 0x3U,
+	.stopInput = CY_TCPWM_INPUT_0,
+	.countInputMode = 0x3U,
+	.countInput = CY_TCPWM_INPUT_1,
+};
+
+static int get_hw_block_info(TCPWM_CNT_Type *reg_addr, cyhal_resource_inst_t *hw_resource)
+{
+	int ret = -EINVAL;
+	uint32_t i;
+
+	for (i = 0u; i < _CYHAL_TCPWM_INSTANCES; i++) {
+		TCPWM_Type *base = _CYHAL_TCPWM_DATA[i].base;
+
+		if (((uint32_t)reg_addr > (uint32_t)base) &&
+		    ((uint32_t)reg_addr < (uint32_t)base + sizeof(TCPWM_Type))) {
```

Comment:
Hey I think that a cleaner way to do this would be to use `uintptr_t` and `POINTER_TO_UINT`, you could initialize `base` as:
```c
uintptr_t base = POINTER_TO_UINT(_CYHAL_TCPWM_DATA[i].base)
```

add a temporary variable for `reg_addr` and I guess `cnt`, and then you would not have to carry castings around the whole function.

Reply:
Simplified the casts around the function